### PR TITLE
#1981 - Request Offering Change: Ministry View [Approve / Deny Request] - Part 3

### PR DESCRIPTION
**Pull request by sh16011993**
## As a part of this PR, the following tasks were completed:

- [x]  **Added an email notification for the scenario: offering change request approved or declined by SABC**: This task involves DB migration to add the new notification type.

Screenshot:

image

##  

- [x]  **e2e tests**

ApplicationOfferingChangeRequestAESTController(e2e)-assessApplicationOfferingChangeRequest
    √ Should assess the application offering change request for the provided application offering change request id.
    √ Should not create a new assessment when the application offering change request has been declined by the ministry user.
    √ Should throw a HttpStatus Not Found (404) error when an application offering change is not in a valid status to be assessed.
    √ Should throw a HttpStatus Bad Request (400) error when an invalid application offering change request status is provided.
    √ Should throw a HttpStatus Forbidden (403) error when an unauthorized AEST user tries to assess the application offering change request.

### CODE REVIEW

File: sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.aest.controller.assessApplicationOfferingChangeRequest.e2e-spec.ts [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+      await saveFakeApplicationOfferingRequestChange(
+        db,
+        {
+          application,
+        },
+        {
+          initialValues: {
+            applicationOfferingChangeRequestStatus:
+              ApplicationOfferingChangeRequestStatus.InProgressWithSABC,
+          } as ApplicationOfferingChangeRequest,
```
Comment (andrepestana-aot) [inline sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.aest.controller.assessApplicationOfferingChangeRequest.e2e-spec.ts:51]: There's no need for the `as ApplicationOfferingChangeRequest` here.

File: sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts [hunk 2]
```diff
@@ -187,6 +188,42 @@ export class NotificationActionsService {
     );
   }
 
+  /**
+   * Creates a notification when an Application Offering Change Request is completed by the ministry.
+   * @param notification input parameters to generate the notification.
+   * @param auditUserId user that should be considered the one that is causing the changes.
+   * @param entityManager optional entity manager to execute in transaction.
+   */
+  async saveApplicationOfferingChangeRequestCompletedByMinistry(
```
Comment (andrepestana-aot) [inline sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts:197]: This name is not related to "create a notification". It sounds like it is saving the offering change request instead.

Comment (andrepestana-aot) [reply sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts:197]: Maybe not for this PR but the method name in line 161 also have this problem (missing "Notification" in the end of the name)

File: sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts [hunk 3]
```diff
@@ -187,6 +188,42 @@ export class NotificationActionsService {
...
+    notification: ApplicationOfferingChangeRequestCompletedByMinistryNotification,
+    auditUserId: number,
+    entityManager?: EntityManager,
```
Comment (andrepestana-aot) [inline sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts:200]: In this method `entityManager` should be mandatory.

**Review by andrepestana-aot — COMMENTED** [review]
Good job. Left some comments. Please take a look.

File: sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.aest.controller.assessApplicationOfferingChangeRequest.e2e-spec.ts [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        statusCode: HttpStatus.NOT_FOUND,
+        message:
+          "Application offering change not found or not in valid status to be approved/declined.",
+        error: "Not Found",
+      });
+  });
+
+  it("Should throw a HttpStatus Bad Request (400) error when an invalid application offering change request status is provided.", async () => {
+    // Arrange
+    const application = await saveFakeApplication(db.dataSource, undefined, {
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.aest.controller.assessApplicationOfferingChangeRequest.e2e-spec.ts:157]: When the bad request is returned from the class validators the endpoint code will not be invoked so there is no need to create fake data.

File: sources/packages/backend/apps/api/src/services/application-offering-change-request/application-offering-change-request.service.ts [hunk 5]
```diff
@@ -605,6 +606,21 @@ export class ApplicationOfferingChangeRequestService {
       await transactionalEntityManager
         .getRepository(ApplicationOfferingChangeRequest)
         .save(applicationOfferingChangeRequest);
+      // Send the application offering change request completed notification.
+      const systemUser = await this.systemUsersService.systemUser();
+      await this.notificationActionsService.saveApplicationOfferingChangeRequestCompletedByMinistry(
+        {
+          givenNames:
+            applicationOfferingChangeRequest.application.student.user.firstName,
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/services/application-offering-change-request/application-offering-change-request.service.ts:614]: As a suggestion, a const can be created to keep the `applicationOfferingChangeRequest.application.student`.

**Review by andrewsignori-aot — COMMENTED** [review]
Nice work, just some minor comments.

File: sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts (hunk 2 above)
Comment (dheepak-aot) [reply sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts:197]: @andrepestana-aot  I believe that must be part of the PR.

File: sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts [hunk 6]
```diff
@@ -187,6 +188,42 @@ export class NotificationActionsService {
...
   }
 
+  /**
+   * Creates a notification when an Application Offering Change Request is completed by the ministry.
+   * @param notification input parameters to generate the notification.
+   * @param auditUserId user that should be considered the one that is causing the changes.
+   * @param entityManager optional entity manager to execute in transaction.
+   */
+  async saveApplicationOfferingChangeRequestCompletedByMinistry(
+    notification: ApplicationOfferingChangeRequestCompletedByMinistryNotification,
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts:198]: Minor: name seems to be very long and can be something like `ApplicationOfferingChangeRequestCompleteNotification`

File: sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts [hunk 7]
```diff
@@ -187,6 +188,42 @@ export class NotificationActionsService {
...
+    entityManager?: EntityManager,
+  ): Promise<void> {
+    const templateId = await this.notificationMessageService.getTemplateId(
+      NotificationMessageType.ApplicationOfferingChangeRequestCompletedByMinistry,
+    );
+    const messagePayload: NotificationEmailMessage = {
+      email_address: notification.toAddress,
+      template_id: templateId,
+      personalisation: {
+        givenNames: notification.givenNames ?? "",
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts:209]: Not a part of this PR, we can have a common filler method in notification scope as this code is repeating for all methods.

File: sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts (hunk 3 above)
Comment (dheepak-aot) [reply sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts:200]: I agree. I also see that all the methods in this class have entityManager mandatory but it is not utilized in an optional way.

File: sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.aest.controller.assessApplicationOfferingChangeRequest.e2e-spec.ts [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+    ).toBe(ApplicationOfferingChangeRequestStatus.Approved);
+    expect(
+      updatedApplicationOfferingChangeRequest.assessedNote.description,
+    ).toBe(note);
+  });
+
+  it("Should throw a HttpStatus Not Found (404) error when an application offering change is assessed for an application not in completed state.", async () => {
+    // Arrange
+    const application = await saveFakeApplication(db.dataSource, undefined, {
+      applicationStatus: ApplicationStatus.Overwritten,
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.aest.controller.assessApplicationOfferingChangeRequest.e2e-spec.ts:89]: If the intention is validate overwritten applications not being processed, I would say line 104 must have a valid status (InProgressWithSABC) to approve.

Btw. I am not sure if this scenario is required.

File: sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.aest.controller.assessApplicationOfferingChangeRequest.e2e-spec.ts [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+        },
+        {
+          initialValues: {
+            applicationOfferingChangeRequestStatus:
+              ApplicationOfferingChangeRequestStatus.InProgressWithSABC,
+          } as ApplicationOfferingChangeRequest,
+        },
+      );
+    const note = "Some dummy note.";
+    const token = await getAESTToken(AESTGroups.BusinessAdministrators);
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.aest.controller.assessApplicationOfferingChangeRequest.e2e-spec.ts:55]: Just wondering should we need a test with non business administrator trying to approve/decline?

**Comment by dheepak-aot** [comment]
Nice work @sh16011993

File: sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts (hunk 2 above)
Comment (sh16011993) [reply sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts:197]: Fixed in both the places.

File: sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts (hunk 6 above)
Comment (sh16011993) [reply sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts:198]: Every other method in the file is having the verb `save` associated. Is it fine if I keep it `saveApplicationOfferingChangeRequestCompleteNotification` ?

File: sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.aest.controller.assessApplicationOfferingChangeRequest.e2e-spec.ts [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+  let applicationOfferingChangeRequestRepo: Repository<ApplicationOfferingChangeRequest>;
+
+  beforeAll(async () => {
+    const { nestApplication, dataSource } = await createTestingAppModule();
+    app = nestApplication;
+    db = createE2EDataSources(dataSource);
+    applicationOfferingChangeRequestRepo = dataSource.getRepository(
+      ApplicationOfferingChangeRequest,
+    );
+  });
```
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.aest.controller.assessApplicationOfferingChangeRequest.e2e-spec.ts:34]: why is this initialization required? we can use `db.<table>` right?

File: sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.aest.controller.assessApplicationOfferingChangeRequest.e2e-spec.ts (hunk 8 above)
Comment (sh16011993) [reply sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.aest.controller.assessApplicationOfferingChangeRequest.e2e-spec.ts:89]: I have removed this scenario @dheepak-aot. Ideally, this scenario shouldn't arise.

File: sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.aest.controller.assessApplicationOfferingChangeRequest.e2e-spec.ts [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+      .send({
+        note,
+        applicationOfferingChangeRequestStatus:
+          ApplicationOfferingChangeRequestStatus.Approved,
+      })
+      .auth(token, BEARER_AUTH_TYPE)
+      .expect(HttpStatus.OK);
+    // Check if the application offering change request was assessed as expected.
+    const updatedApplicationOfferingChangeRequest =
+      await applicationOfferingChangeRequestRepo.findOne({
```
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.aest.controller.assessApplicationOfferingChangeRequest.e2e-spec.ts:69]: how about `await db.applicationOfferingChangeRequest.findOne()`,?

File: sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.aest.controller.assessApplicationOfferingChangeRequest.e2e-spec.ts [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+        applicationOfferingChangeRequestStatus:
+          ApplicationOfferingChangeRequestStatus.Approved,
+      })
+      .auth(token, BEARER_AUTH_TYPE)
+      .expect(HttpStatus.OK);
+    // Check if the application offering change request was assessed as expected.
+    const updatedApplicationOfferingChangeRequest =
+      await db.applicationOfferingChangeRequest.findOne({
+        select: {
+          id: true,
```
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.aest.controller.assessApplicationOfferingChangeRequest.e2e-spec.ts:65]: looks like id here and in line 67 is not used , do we need it?

File: sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.aest.controller.assessApplicationOfferingChangeRequest.e2e-spec.ts [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+          applicationOfferingChangeRequestStatus: true,
+          assessedNote: { id: true, description: true },
+        },
+        relations: { assessedNote: true },
+        where: { id: applicationOfferingChangeRequest.id },
+      });
+    expect(
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.aest.controller.assessApplicationOfferingChangeRequest.e2e-spec.ts:72]: Just added a similar comment to another PR, we should always try to validate all the impacted DB columns as possible.
For instance, the modifier can also be checked here and I missed on previous reviews.

File: sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.aest.controller.assessApplicationOfferingChangeRequest.e2e-spec.ts (hunk 12 above)
Comment (sh16011993) [reply sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.aest.controller.assessApplicationOfferingChangeRequest.e2e-spec.ts:65]: IMU, this `id` is required by the relations for doing the join. I have removed the `id` from line 67.

File: sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.aest.controller.assessApplicationOfferingChangeRequest.e2e-spec.ts [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+
+  it("Should throw a HttpStatus Bad Request (400) error when an invalid application offering change request status is provided.", async () => {
+    // Arrange
+    const note = "Some dummy note.";
+    const token = await getAESTToken(AESTGroups.BusinessAdministrators);
+    const endpoint = "/aest/application-offering-change-request/9999";
+    // Act/Assert
+    await request(app.getHttpServer())
+      .patch(endpoint)
+      .send({
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.aest.controller.assessApplicationOfferingChangeRequest.e2e-spec.ts:127]: I am not sure how the others E2E are doing it but the payload belongs more to the `Arrange` section.

**Review by andrewsignori-aot — APPROVED** [review]
Looks good 👍

File: sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.aest.controller.assessApplicationOfferingChangeRequest.e2e-spec.ts (hunk 12 above)
Comment (ann-aot) [reply sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.aest.controller.assessApplicationOfferingChangeRequest.e2e-spec.ts:65]: I am sorry, why id is required for join? I am not following.

Comment (andrepestana-aot) [reply sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.aest.controller.assessApplicationOfferingChangeRequest.e2e-spec.ts:65]: Yes, I checked that and I had an error without selecting the id. I had a similar issue with another `findOne` query.

Comment (ann-aot) [reply sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.aest.controller.assessApplicationOfferingChangeRequest.e2e-spec.ts:65]: oh..okay, but when I checked the code, we have some type typeorm `findOne` query with `relation` without `id`. Now I suspect those query working or not

**Review by ann-aot — APPROVED** [review]
LGTM 👍 But added a reply for a comment https://github.com/bcgov/SIMS/pull/2422#discussion_r1362898388, if there is an error, we might need to create a ticket to revisit/validate those queries in our code base.

**Review by dheepak-aot — APPROVED** [review]
Thanks for doing the changes 👍

File: sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.aest.controller.assessApplicationOfferingChangeRequest.e2e-spec.ts (hunk 12 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/route-controllers/application-offering-change-request/_tests_/e2e/application-offering-change-request.aest.controller.assessApplicationOfferingChangeRequest.e2e-spec.ts:67]: @ann-aot I saw the missing ID issue happening yesterday while @andrepestana-aot was sharing the screen but the queries that you provided prove that it can work.
It may be a not-so-evident error or something that we are missing in the entity model classes.

**Review by andrewsignori-aot — APPROVED** [review]
Thanks for doing the changes, looks good 👍

**Review by andrepestana-aot — APPROVED** [review]
Looks good! Thanks for the changes!
